runtime/src/test/java/com/dylibso/chicory/runtime/HostImportsTest.java

```diff
@@ -177,5 +181,82 @@ void addIndex() {
                 assertEquals(1, result.index().length);
             }
         }
+
+        @Nested
+        class DSL {
+
+            @Test
+            void withIndex() {
+                var moduleName = "module";
+                var fieldName = "filed";
+                HostImports.builder()
+                        .withNewImport(moduleName, fieldName)
```

Comment:
I find this style of nested builder to be strange, especially since only one method can be called. For nested builders, I've found that having a `Consumer` argument is a good approach, but that doesn't seem useful here. The only thing we are saving by having `withNewImport()` is the two name arguments.

All of the `with` methods on `Builder` replace the entire collection. For example, `withGlobals()` replaces all of the globals, whereas `addGlobal()` adds additional globals. So the name `withNewImport()` seems inconsistent.

Instead, what if we put new methods directly on `Builder`:
```java
.addGlobal(moduleName, fieldName, Value.i32(1))
.addGlobal(moduleName, fieldName, MutabilityType.Var, Value.i32(1))
.addMutableGlobal(moduleName, fieldName, Value.i32(1))
.addMemory(moduleName, fieldName)
.addMemory(moduleName, fieldName, new MemoryLimits(1))
.addMemory(moduleName, fieldName, 1) // I'd drop these as the MemoryLimits version seems more clear
.addMemory(moduleName, fieldName, 1, 2)
.addTable(moduleName, fieldName)
.addTable(moduleName, fieldName, ValueType.ExternRef)
.addTable(moduleName, fieldName, new Limits(1))
.addTable(moduleName, fieldName, 1) // same, I'd drop these and use the Limits version
.addTable(moduleName, fieldName, 1, 2)
// does type inference work out if we name all of these "addFunction"?
.addProcedure(moduleName, fieldName, () -> System.out.println("hello world"))
.addProcedure(moduleName, fieldName, (Instance inst) -> () -> System.out.println("hello world"))
.addSupplier(moduleName, fieldName, () -> 1)

Reply:
Thanks for your input @electrum !
A few different subjects:

- `add` instead of `with`: agree will change accordingly, thanks for noticing!
- `two name arguments`: personally, I feel it awkward to have to type those "names" along with the implementation, the nested builder was an attempt to fix it.
- `new methods directly on Builder`: this works for sure, I'm afraid is an improvement less impactful than I originally thought

Reply:
> I feel it awkward to have to type those "names" along with the implementation

This test might look worse due to repeating the same variables. In actual usage, the field names will need to be unique (and the builder should validate that)? So if we change it to use constants, does it look better?

```java
.addGlobal("test", "g1", Value.i32(1))
.addGlobal("test", "g2", MutabilityType.Var, Value.i32(1))
.addMutableGlobal("test", "gmut", Value.i32(1))
.addMemory("test", "m1")
.addMemory("test", "mlim", new MemoryLimits(1))
.addTable("test", "t1")
.addTable("test", "tref", ValueType.ExternRef)
.addTable("test", "t3", new Limits(1))
```

Another problem is that the code formatter forces each chained call to be on a separate line. If we were formatting the code by hand, it would look better:
```java
.add("test", "g1").global(Value.i32(1))
.add("test", "g2").global(MutabilityType.Var, Value.i32(1))
.add("test", "gmut").mutableGlobal(Value.i32(1))
.add("test", "m1").memory()
.add("test", "mlim").memory(new MemoryLimits(1))
.add("test", "t1").table()
.add("test", "tref").table(ValueType.ExternRef)
.add("test", "t3").table(new Limits(1))
```
Which I agree looks slightly cleaner, but we can't format like that...

Reply:
I love this!

```java
.add("test", "g1").global(Value.i32(1))
```
I believe that we should not be limited by our own formatter, and we can always disable it when necessary :smirk: